### ISSUE TITLE
Number of plot ticks on axis should scale based on available space

### ISSUE THREAD

**Issue by VWoeltjen** [nasa/openmct#620]
<p>In plots, particularly currently on the Y axis, the number of ticks that are displayed should dynamically update based on the space available to display them. To illustrate, shrink the size of a plot object to be quite small and observe plot tick rendering on the Y axis. </p>

<p>I propose the algorithm for this go something like:</p>

<p>&lt;# of ticks&gt; = integer(&lt;available px on axis&gt; / &lt;px needed per tick&gt;)</p>

<p>On the Y axis, &lt;px needed per tick&gt; is probably in the neighborhood of 20px or so. </p>

<p>The X axis uses a different strategy right now (the number of ticks is quite low) so it doesn't need as much attention.</p>

Migrated from: https://trunk.arc.nasa.gov/jira/browse/WTD-451
Originally created: Tue, 30 Sep 2014 11:40:36 -0700
Updated: Thu, 21 Jan 2016 15:17:54 -0800

**Comment by hudsonfoo** [comment]
@VWoeltjen any thoughts about this CSS only solution? http://quick.as/xAQfnQBp

```diff
		&.gl-plot-y {
+			@media screen and (max-height:300px){
+				.gl-plot-tick:nth-child(2n+1) {
+					display: none;
+				}
+			}
			top: $plotLegendH + $interiorMargin;
			right: auto;
			bottom: nth($plotDisplayArea, 3);
			left: 0;
			width: $plotYBarW;
		}
```

While it doesn't use the equation you mentioned, this solution maintains the separation of the view from the controller, and I think it could be tinkered with to get it a lot closer to what you're wanting. There's probably a better way to do this that @charlesh88 can think up as well.

**Comment by hudsonfoo** [comment]
Adding a little fade effect to the ticks actually looks really nice as well.

**Comment by charlesh88** [comment]
I'm a little confused: given 7 ticks, 2n+0 should exhibit the effect you show in the screen capture, which is closer to what we want. 2n+1 (shown in the code snippet) hides the odd number ticks, which would hide the first and last ticks, which would be bad as they are the upper and lower limits of the plot area. Also, some plots use string labels on their Y axis; simply removing ticks wouldn't be a good solution for those.

Media queries only help us with regard to screen size, not parent container size. We really don't have a problem currently with ticks in the main view area; the situation we really need to handle is when a plot is in a Layout and is sized small there. I really wish we could have a CSS interrogator for the size of a parent container, but the W3C isn't answering my calls on that one ;)

Bottomline, I think this is fine, but I think we really should look at furthering our D3 integration here to allow the ticks to be calced dynamically based on the size of the given object, in the way that we've done it in the new Time Conductor time axis element.

**Comment by hudsonfoo** [comment]
Totally agree about media queries. I figured I would find a  way to overcome that if you guys were good with a CSS route. That being said, I'm definitely down to figure this out with D3.

Not sure what the deal was with 2n+1. I did a straight copy from my little test into the comment, but I think you're right, that shouldn't have worked.

**Comment by VWoeltjen** [comment]
For purposes of linkage, #977 is an umbrella issue for improving tick-handling generally (with links to a few other sub-issues) and could be a useful point of reference here.

In particular, we should avoid solutions here that will be incompatible with or complicated by solutions to those related issues. That's also a significant reason for preferring something like [d3-axis](https://github.com/d3/d3-axis) here; it looks like it meets (or should be compatible with) those requirements already.

**Comment by charlesh88** [comment]
Still need this!

**Comment by larkin** [comment]
Implemented in #1557, though it's possible to improve upon that implementation.  Going to close this for now and we can reopen after we get more experience with new plots.
